836 public search option

tcnichol opened the pull request:
This pull request will enable a search of public data for non-logged in users.

It also fixes a problem. When a logged in user would search, public and authenticated matches were not being shown. In the query 'public' and 'authenticated' had to be in all lowercase. Additionally, not all files were having their status updated when the dataset changed. This is also fixed.

To test, create datasets. Make sure some are public and authenticated.

If no user is logged in, public datasets and files should show up in search results. 

If a user is logged in, public and authenticated datasets and files should show up in search results, along with their own items or datasets explicitly shared. 

Changing the status of a dataset should be reflected in search results.

ddey2 reviewed (changes requested):
@tcnichol I can see the public datasets in the dropdown list for public search. But when I click enter or click on any of the dataset form the dropdown list, it doesn't take me to the search result page or the dataset page itself.
<img width="1116" alt="Screenshot 2024-03-11 at 3 31 35 PM" src="https://github.com/clowder-framework/clowder2/assets/22405928/93d259e5-8e89-4a24-9862-2101d9201fb7">

tcnichol commented:
> @tcnichol I can see the public datasets in the dropdown list for public search. But when I click enter or click on any of the dataset form the dropdown list, it doesn't take me to the search result page or the dataset page itself. <img alt="Screenshot 2024-03-11 at 3 31 35 PM" width="1116" src="https://private-user-images.githubusercontent.com/22405928/311858972-93d259e5-8e89-4a24-9862-2101d9201fb7.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTAyNjM5MjYsIm5iZiI6MTcxMDI2MzYyNiwicGF0aCI6Ii8yMjQwNTkyOC8zMTE4NTg5NzItOTNkMjU5ZTUtOGU4OS00YTI0LTk4NjItMjEwMWQ5MjAxZmI3LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAzMTIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMzEyVDE3MTM0NlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTgyMGRjZDA2YWUzMzExY2RhMzdmZTVhYTQyOGZhNDNmODMzYTVkYmJiOWM4ZTJhNDczOWYwYTdjNWZhZDEzMWImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.ufuJPw99rd2TmCGTf9rNspZLeBhQTELYBFDKTrzwpBg">

I think I have fixed this. I noticed that in the search results the links were not for public_datasets or public_files, they were the same as for logged in search. I added a PublicSearchResult component which should fix these issues.

ddey2 commented on frontend/src/components/search/PublicSearch.tsx:
```diff
@@ -0,0 +1,210 @@
+							label={
+								<Typography
+									variant="body1"
+									sx={{
```
Can you put the styles in main.css, may be?

ddey2 commented on frontend/src/components/search/PublicSearchResult.tsx:
```diff
@@ -0,0 +1,104 @@
+		return (
+			<List
+				sx={{
+					width: "100%",
```
here too

ddey2 commented on frontend/src/components/search/PublicSearchResult.tsx:
```diff
@@ -0,0 +1,104 @@
+				<MuiLink
+					component={Link}
+					to={`/public_files/${item._id}?public_dataset=${item.dataset_id}`}
+					sx={{ fontWeight: "bold", fontSize: "18px" }}
```
here too.

ddey2 commented on frontend/src/components/search/PublicSearchResult.tsx:
```diff
@@ -0,0 +1,104 @@
+				<MuiLink
+					component={Link}
+					to={`/public_datasets/${item._id}`}
+					sx={{ fontWeight: "bold", fontSize: "18px" }}
```
here too...looks like you're using the same styling properties again.

tcnichol replied:
I created a separate issue for these.

tcnichol commented:
@longshuicy 

I checked and the search result links should be fixed for public search now, for both files and datasets.

ddey2 commented on frontend/src/components/PublicLayout.tsx:
```diff
@@ -235,6 +235,18 @@ export default function PersistentDrawerLeft(props) {
+					<ListItem key={"public_search"} disablePadding>
+						<ListItemButton component={RouterLink} to="/public_search">
+							<ListItemIcon>
+								<SearchDatasetIcon />
```
It gives error here as you never imported SearchDatasetIcon in this file.

longshuicy replied:
@tcnichol is this addressed?

tcnichol replied:
Just pushed a fix for this.

ddey2 reviewed (changes requested):
Rest looks good to me. Once you import SearchDatasetIcon, public search works.

ddey2 commented:
@tcnichol Clicking on the link after the search doesn't work. Clicking on the dataset takes you to the main page.
<img width="1651" alt="Screenshot 2024-04-16 at 2 39 53 PM" src="https://github.com/clowder-framework/clowder2/assets/22405928/1ba0789a-c4cc-4520-9d81-392c8ed74fe2">

ddey2 reviewed (changes requested):
left comment

tcnichol commented:
@ddey2 I noticed that there was an inconsistency in the routes for public_datasets. Some spots had public/datasets instead. Went through and I think I fixed that.

Though now the link to the public search is broken at the top of the Public page, working on that.

ddey2 reviewed (approved):
Looks good.

longshuicy commented on backend/app/routers/elasticsearch.py:
```diff
@@ -23,8 +23,8 @@ def _add_permissions_clause(
-                {"term": {"status": "AUTHENTICATED"}},
-                {"term": {"status": "PUBLIC"}},
+                {"term": {"status": "authenticated"}},
+                {"term": {"status": "public"}},
```
Just curious, why capitalized doesn't work? In the database and in the router logic, it's all capitalized?

tcnichol replied:
I have fixed the problem with the files having null status. They now should always have the same status as the dataset.

I'm still not getting files in the search results for public search, looking into this.

longshuicy replied:
I don't think this has been resolved yet?

tcnichol replied:
I think this is fixed. Files should be showing up in public search now.

longshuicy commented on frontend/src/routes.tsx:
```diff
@@ -194,7 +195,7 @@ export const AppRoutes = (): JSX.Element => {
 					}
 				/>
-				<Route path="/public/files/:fileId" element={<PublicFileComponent />} />
+				<Route path="/public_files/:fileId" element={<PublicFileComponent />} />
```
Are we using public/files or public_files? 
In other PRs that got merged in, I fixed places that are using "_"...

longshuicy replied:
If you would like to use "_", you might have to 1. merge the latest main, 2. do a project wide search of places still using /
E.g. If you go to a file under public dataset, it right now still goes to /public/file and hence broken.

tcnichol replied:
I think I got all of these fixed with a merge and a search/replace. Testing.

tcnichol replied:
I also think that files should have a status. Checking the populate fake data, the status for files is 'null' in some cases and those should match the datasets.

longshuicy replied:
Thank you. This has been resolved.

longshuicy reviewed (changes requested):
Public search seems to be able to match dataset, not files though. Might be your clause is specifically matching dataset status. Not sure how to get around, but at least I would change the label `Search for Datasets and Files`
<img width="1166" alt="image" src="https://github.com/clowder-framework/clowder2/assets/13950475/71cb9c64-f77e-42f0-8a82-f117a796377f">

longshuicy commented:
@tcnichol any chance you have addressed public files not searchable issue? If it is challenging, could you create another issue to address that separately? The rest of this PR looks good.

tcnichol commented:
> @tcnichol any chance you have addressed public files not searchable issue? If it is challenging, could you create another issue to address that separately? The rest of this PR looks good.

Let me check. I think it should be fixed because the problem was that the fields for public or authenticated were null for files in elasticsearch, but that should be fixed now. I'll make sure it is working.

lmarini commented:
> > @tcnichol any chance you have addressed public files not searchable issue? If it is challenging, could you create another issue to address that separately? The rest of this PR looks good.
> 
> Let me check. I think it should be fixed because the problem was that the fields for public or authenticated were null for files in elasticsearch, but that should be fixed now. I'll make sure it is working.

@tcnichol is this fixed now? Thanks.

longshuicy reviewed (approved):
After drop volume and restart, elasticsearch is indexing properly and I can search for files. Approved.
